neurodamus/utils/memory.py

```diff
@@ -160,6 +162,136 @@ def pretty_printing_memory_mb(memory_mb):
+        if rank_id not in rank_allocation[pop]:
+            rank_allocation[pop][rank_id] = []
+        rank_allocation[pop][rank_id].append(cell_id)
+        # Update the total memory load of the rank
+        total_memory += batch_memory
+        rank_memory[pop][rank_id] = total_memory
+        # Update total memory and re-add to the heap
+        heapq.heappush(ranks, (total_memory, rank_id))
+
+    # Start distributing cells across ranks
+    for pop, metype_gids in dry_run_stats.metype_gids.items():
+        logging.info("Distributing cells of population %s", pop)
+        batch = []
+        batch_memory = 0
+
+        for cell_id, memory in generate_cells(metype_gids):
+            batch.append(cell_id)
+            batch_memory += memory
+            if len(batch) == batch_size:
+                assign_cells_to_rank(batch_memory, pop)
+                batch = []
+                batch_memory = 0
+
+        # Assign any remaining cells in the last, potentially incomplete batch
+        if batch:
+            assign_cells_to_rank(batch_memory, pop)
+
+    return rank_allocation, rank_memory
+
+
+@run_only_rank0
+def print_allocation_stats(rank_allocation, rank_memory):
+    """
+    Print statistics of the memory allocation across ranks.
+
+    Args:
+        rank_allocation (dict): A dictionary where keys are rank IDs and
+                                values are lists of cell IDs assigned to each rank.
+        rank_memory (dict): A dictionary where keys are rank IDs
+                            and values are the total memory load on each rank.
+    """
+    logging.debug("Rank allocation: {}".format(rank_allocation))
+    logging.debug("Total memory per rank: {}".format(rank_memory))
+    import statistics
+    for pop, rank_dict in rank_memory.items():
+        values = list(rank_dict.values())
+        logging.info("Population: {}".format(pop))
+        logging.info("Mean allocation per rank [KB]: {}".format(round(statistics.mean(values))))
+        try:
+            stdev = round(statistics.stdev(values))
+        except statistics.StatisticsError:
+            stdev = 0
+        logging.info("Stdev of allocation per rank [KB]: {}".format(stdev))
+
+
+@run_only_rank0
+def export_allocation_stats(rank_allocation, filename):
+    """
+    Export allocation dictionary to serialized pickle file.
+    """
+    import pickle
+    try:
+        with open(filename, 'wb') as f:
+            pickle.dump(rank_allocation, f)
+    except Exception as e:
+        logging.warning("Unable to export allocation stats: {}".format(e))
+
+
+@run_only_rank0
+def import_allocation_stats(filename):
+    """
+    Import allocation dictionary from serialized pickle file.
+    """
+    import pickle
+    try:
+        with open(filename, 'rb') as f:
+            rank_allocation = pickle.load(f)
+        return rank_allocation
+    except Exception as e:
+        logging.warning("Unable to import allocation stats: {}".format(e))
+        return None
```

Comment:
I'm a bit concerned that this will be hiding errors and hard to debug in the future. Can you please

1. Not catch any exceptions
2. Use JSON instead of `pickle` to be able to inspect the stored data easier?

Reply:
We discussed with @ferdonline about the possibility of using JSON also for this purpose (we are already using it to store the cell memory usage) but then we settled for `pickle` since the allocation table can get very big and we're not expecting the user to view or change the resulting file.

Are there particular usecases where you see the user changing the allocation files? Or any other reason to use JSON instead of `pickle`?

Reply:
Agree with not handling the Exception here. Maybe we should have some API to check if there is load-balancing data. We may want to have it optional or mandatory controlled by another routine with more context.

Reply:
JSON would make it way easier to write test cases. Also if someone wants to debug something, it'll be easier to handle.

Space and time here are not really that significant, IMO. I ran a small test case with 10 million gids spread over 100 types + 100 ranks. If we're concerned about space, it would be _very_ sensible to gzip everything, since even pickle files are sort of just plain text:
```
❯ ls -lh jayson.json{,.gz} rick.pkl{,.gz}
-rw-r--r-- 1 matwolf bbp  38M Feb  7 15:21 jayson.json
-rw-r--r-- 1 matwolf bbp 249K Feb  7 15:20 jayson.json.gz
-rw-r--r-- 1 matwolf bbp  27M Feb  7 15:22 rick.pkl
-rw-r--r-- 1 matwolf bbp 187K Feb  7 15:06 rick.pkl.gz
```

Using the `ujson` library rather than the built-in `json`, times are also pretty similar:
```
❯ ipython test.ipy
JSON write
507 ms ± 1.22 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
JSON read
562 ms ± 493 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
🥒 write
183 ms ± 1.41 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
🥒 read
427 ms ± 13.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
compared to the rest of the setup steps of the simulation, this is all insignificant.

Besides, we don't read the data back in, why write it in the first place? That would reduce the confusing amount of files we scatter everywhere…

Reply:
Thanks for the measurements, I was expecting numpy pickling to be wiser, maybe something is off?

> Besides, we don't read the data back in

That's the big point here. We will read the data back in, and we want to read it back to numpy arrays and avoid lists. Without pickle can we do that? Honeslty it's nonsense to write complex import function or use third party libs

> JSON would make it way easier to write test cases.

I'm not that convinced. Tests can create the structure easily and deserialization is trivial

Reply:
> EDIT: Just noticed this is using plain lists. We should try to use a compact data structure

`libsonata.Selection` compresses nicely if gids are continuous. Otherwise, just running gzip on the files seems to compress superbly. And really, if we run on TB sized circuits, 50 MB is nothing.

> We will read the data back in

When? Until that's happening, we should really hold off a bit on dumping more files.

The very minimal expectation should also be that the file format ends in `.pkl` for pickled files. `.bin` could be anything and will be confusing.

Reply:
Following yesterday round of fixes, we've discovered an issue that prevented some of the allocations to be actually pickled. So, even using a more compact data structure (at the moment the `Vector` structure integrated in `neurodamus` which already helped reducing the file size by a factor of 4) the size of the allocation file is ~200 MB for a 50M circuit.

I will experiment with `libsonata.Selection` (as you suggested somewhere else) and `gzip` compression to see if I can improve this.

Reply:
I've included automatic `gzip` compression so the pickled allocation file for the 50M circuit is down to a more reasonable ~100 MB at basically no performance cost.